In [1]:
# Import the required library
import pandas as pd

# Define the URL of the Excel file on GitHub
url = 'https://raw.githubusercontent.com/lk-learner/IBM-data-analyst-professional-certificate/main/Course-7%20%3A%20Data%20Analysis%20with%20Python/Activities/adult.csv'

# Load the Excel file into a Pandas dataframe
df = pd.read_csv(url)

# Display the dataframe
df.head()



age          workclass  fnlwgt   education  educational-num  \
0   39          State-gov   77516   Bachelors               13   
1   50   Self-emp-not-inc   83311   Bachelors               13   
2   38            Private  215646     HS-grad                9   
3   53            Private  234721        11th                7   
4   28            Private  338409   Bachelors               13   

        marital-status          occupation    relationship    race   gender  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [2]:
df.isna().sum()

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64

In [3]:
df.size

488415

In [4]:
df.shape

(32561, 15)

In [5]:
df.dtypes

age                 int64
workclass          object
fnlwgt              int64
education          object
educational-num     int64
marital-status     object
occupation         object
relationship       object
race               object
gender             object
capital-gain        int64
capital-loss        int64
hours-per-week      int64
native-country     object
income             object
dtype: object

In [6]:
df['workclass'].unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay',
       ' Never-worked'], dtype=object)

In [7]:
df['education'].unique()

array([' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th',
       ' Some-college', ' Assoc-acdm', ' Assoc-voc', ' 7th-8th',
       ' Doctorate', ' Prof-school', ' 5th-6th', ' 10th', ' 1st-4th',
       ' Preschool', ' 12th'], dtype=object)

In [8]:
df['marital-status'].unique()

array([' Never-married', ' Married-civ-spouse', ' Divorced',
       ' Married-spouse-absent', ' Separated', ' Married-AF-spouse',
       ' Widowed'], dtype=object)

In [9]:
df['occupation'].unique()

array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Sales', ' Craft-repair',
       ' Transport-moving', ' Farming-fishing', ' Machine-op-inspct',
       ' Tech-support', ' ?', ' Protective-serv', ' Armed-Forces',
       ' Priv-house-serv'], dtype=object)

In [10]:
df['relationship'].unique()

array([' Not-in-family', ' Husband', ' Wife', ' Own-child', ' Unmarried',
       ' Other-relative'], dtype=object)

In [11]:
df['race'].unique()

array([' White', ' Black', ' Asian-Pac-Islander', ' Amer-Indian-Eskimo',
       ' Other'], dtype=object)

In [12]:
df['gender'].unique()

array([' Male', ' Female'], dtype=object)

In [13]:
df['native-country'].unique()

array([' United-States', ' Cuba', ' Jamaica', ' India', ' ?', ' Mexico',
       ' South', ' Puerto-Rico', ' Honduras', ' England', ' Canada',
       ' Germany', ' Iran', ' Philippines', ' Italy', ' Poland',
       ' Columbia', ' Cambodia', ' Thailand', ' Ecuador', ' Laos',
       ' Taiwan', ' Haiti', ' Portugal', ' Dominican-Republic',
       ' El-Salvador', ' France', ' Guatemala', ' China', ' Japan',
       ' Yugoslavia', ' Peru', ' Outlying-US(Guam-USVI-etc)', ' Scotland',
       ' Trinadad&Tobago', ' Greece', ' Nicaragua', ' Vietnam', ' Hong',
       ' Ireland', ' Hungary', ' Holand-Netherlands'], dtype=object)

3 columns have '?' values. They are 'Native-country','occupation','workclass'

Since we have dataset of 32561 values. Lets remove rows with '?' values

In [14]:
# Replace '?' with NaN
import numpy as np
df.replace(' ?', np.nan, inplace=True)

# Drop rows with missing values
df.dropna(inplace=True)

# Print the number of rows and columns in the cleaned dataset
print(f"The cleaned dataset has {df.shape[0]} rows and {df.shape[1]} columns.")

The cleaned dataset has 30162 rows and 15 columns.


In [15]:
df['occupation'].unique()

array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Sales', ' Transport-moving',
       ' Farming-fishing', ' Machine-op-inspct', ' Tech-support',
       ' Craft-repair', ' Protective-serv', ' Armed-Forces',
       ' Priv-house-serv'], dtype=object)

In [16]:
df['workclass'].unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' Self-emp-inc', ' Without-pay'], dtype=object)

In [17]:
df['native-country'].unique()

array([' United-States', ' Cuba', ' Jamaica', ' India', ' Mexico',
       ' Puerto-Rico', ' Honduras', ' England', ' Canada', ' Germany',
       ' Iran', ' Philippines', ' Poland', ' Columbia', ' Cambodia',
       ' Thailand', ' Ecuador', ' Laos', ' Taiwan', ' Haiti', ' Portugal',
       ' Dominican-Republic', ' El-Salvador', ' France', ' Guatemala',
       ' Italy', ' China', ' South', ' Japan', ' Yugoslavia', ' Peru',
       ' Outlying-US(Guam-USVI-etc)', ' Scotland', ' Trinadad&Tobago',
       ' Greece', ' Nicaragua', ' Vietnam', ' Hong', ' Ireland',
       ' Hungary', ' Holand-Netherlands'], dtype=object)

In [18]:
df.shape

(30162, 15)

In [19]:
#Cleaning column values 
col = ['workclass', 'education', 'marital-status', 'occupation', 'relationship','race','gender','native-country','income']
for i in col:
    df[i] = df[i].str.strip()

In [20]:
# # Convert object columns to string
# for col in df.select_dtypes(include='object').columns:
#     df[col] = df[col].astype(str)

# # Print the data types of each column
# print(df.dtypes)

#### 1. How many men and women (sex feature) are represented in this dataset?


In [21]:
df['income'].unique()

array(['<=50K', '>50K'], dtype=object)

In [22]:
# count the number of men and women
gender_counts = df['gender'].value_counts()
print(gender_counts)

Male      20380
Female     9782
Name: gender, dtype: int64


#### 2. What is the average age (age feature) of women?

In [23]:
# Calculate the average age of women
avg_age_women = df[df['gender'] == 'Female']['age'].mean()

# Print the result
print(f"The average age of women is {round(avg_age_women,2)} years old.")

The average age of women is 36.88 years old.


#### 3. What is the percentage of German citizens (native-country feature)?


In [24]:
# Calculate the percentage of German citizens
german_citizens = df[df['native-country'] == 'Germany']
percentage_german_citizens = (len(german_citizens) / len(df)) * 100

# Print the result
print(f"The percentage of German citizens is {round(percentage_german_citizens, 2)}%.")

The percentage of German citizens is 0.42%.


#### 4. What are the mean and standard deviation of age for those who earn more than 50K per year (salary feature) and those who earn less than 50K per year?


In [25]:
df['income'].unique()

array(['<=50K', '>50K'], dtype=object)

In [26]:
# mean and standard deviation of age for those who earn more than 50K per year
high_income = df[df['income'] == '>50K']
high_income_age_mean = high_income['age'].mean()
high_income_age_std = high_income['age'].std()

print("Mean age of those who earn more than 50K per year:", high_income_age_mean)
print("Standard deviation of age of those who earn more than 50K per year:", high_income_age_std)

# mean and standard deviation of age for those who earn less than or equal to 50K per year
low_income = df[df['income'] == '<=50K']
low_income_age_mean = low_income['age'].mean()
low_income_age_std = low_income['age'].std()

print("Mean age of those who earn less than or equal to 50K per year:", low_income_age_mean)
print("Standard deviation of age of those who earn less than or equal to 50K per year:", low_income_age_std)


Mean age of those who earn more than 50K per year: 43.95911028236548
Standard deviation of age of those who earn more than 50K per year: 10.269632835673852
Mean age of those who earn less than or equal to 50K per year: 36.60806038668668
Standard deviation of age of those who earn less than or equal to 50K per year: 13.464631257161633


#### 5.  Is it true that people who earn more than 50K have at least a high school education? (education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters or Doctorate feature)


In [27]:
high_income_education = df[df['income'] == '>50K']['education'].value_counts()
print(high_income_education)


Bachelors       2126
HS-grad         1617
Some-college    1336
Masters          918
Prof-school      406
Assoc-voc        344
Doctorate        280
Assoc-acdm       256
10th              59
11th              59
7th-8th           35
12th              29
9th               25
5th-6th           12
1st-4th            6
Name: education, dtype: int64


#### 6. Display age statistics for each race (race feature) and each gender (sex feature). Use groupby() and describe(). Find the maximum age of men of Amer-Indian-Eskimo race.

In [28]:
# age statistics for each race
race_stats = df.groupby('race')['age'].describe()
print(race_stats)

# age statistics for each gender
gender_stats = df.groupby('gender')['age'].describe()
print(gender_stats)


                      count       mean        std   min   25%   50%    75%  \
race                                                                         
Amer-Indian-Eskimo    286.0  36.737762  11.630168  17.0  28.0  35.0  44.75   
Asian-Pac-Islander    895.0  37.910615  12.587717  17.0  28.0  36.0  46.00   
Black                2817.0  37.860845  12.418791  17.0  28.0  37.0  46.00   
Other                 231.0  33.601732  10.949163  17.0  26.0  31.0  41.00   
White               25933.0  38.580612  13.250943  17.0  28.0  37.0  47.00   

                     max  
race                      
Amer-Indian-Eskimo  80.0  
Asian-Pac-Islander  90.0  
Black               90.0  
Other               77.0  
White               90.0  
          count       mean        std   min    25%   50%   75%   max
gender                                                              
Female   9782.0  36.883459  13.532427  17.0  25.25  35.0  46.0  90.0
Male    20380.0  39.184004  12.873243  17.0  29.00  38.0 

#### 7. Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (marital-status feature)? Consider as married those who have a marital-status starting with Married (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.


In [29]:
# create a new column to identify marital status
df['marital_status'] = df['marital-status'].apply(lambda x: 'Married' if x.startswith('Married') else 'Single')

# proportion of those who earn more than 50K among married and single men
proportion = df[(df['gender'] == 'Male') & (df['marital_status'].isin(['Married', 'Single']))].groupby(['marital_status', 'income'])['income'].count() / df[(df['gender'] == 'Male') & (df['marital_status'].isin(['Married', 'Single']))].groupby(['marital_status'])['income'].count()

print(proportion)


marital_status  income
Married         <=50K     0.552016
                >50K      0.447984
Single          <=50K     0.911506
                >50K      0.088494
Name: income, dtype: float64


#### 8. What is the maximum number of hours a person works per week (hours-per-week feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?


In [30]:
# maximum number of hours a person works per week
max_hours = df['hours-per-week'].max()
print("Maximum number of hours a person works per week:", max_hours)


Maximum number of hours a person works per week: 99


In [31]:
# number of people who work the maximum number of hours
num_people = df[df['hours-per-week'] == max_hours]['hours-per-week'].count()
print("Number of people who work the maximum number of hours:", num_people)


Number of people who work the maximum number of hours: 78


In [32]:
# percentage of those who earn more than 50K among those who work the maximum number of hours
num_high_income = df[(df['hours-per-week'] == max_hours) & (df['income'] == '>50K')]['income'].count()
percentage = (num_high_income / num_people) * 100
print("Percentage of those who earn more than 50K among those who work the maximum number of hours:", percentage, "%")


Percentage of those who earn more than 50K among those who work the maximum number of hours: 32.05128205128205 %


#### 9. Count the average time of work (hours-per-week) for those who earn a little and a lot (salary) for each country (native-country). What will these be for Japan?


In [33]:
# average time of work for each country and salary group
avg_work_time = df.groupby(['native-country', 'income'])['hours-per-week'].mean().reset_index()
avg_work_time


native-country income  hours-per-week
0        Cambodia  <=50K       41.545455
1        Cambodia   >50K       40.000000
2          Canada  <=50K       38.704225
3          Canada   >50K       46.944444
4           China  <=50K       38.541667
..            ...    ...             ...
75  United-States   >50K       45.750536
76        Vietnam  <=50K       38.152542
77        Vietnam   >50K       39.200000
78     Yugoslavia  <=50K       41.600000
79     Yugoslavia   >50K       49.500000

[80 rows x 3 columns]

In [34]:
# average time of work for each salary group in Japan
japan_data = df.loc[df['native-country'] == 'Japan']
japan_avg_work_time = japan_data.groupby('income')['hours-per-week'].mean()

print(japan_avg_work_time)


income
<=50K    41.333333
>50K     48.086957
Name: hours-per-week, dtype: float64


#### 10. Find out the total number of hours worked and mean salary as per different occupations.
                           

In [35]:
total_num_hrs_occupations = df.groupby(['native-country'])['hours-per-week'].sum().reset_index()
total_num_hrs_occupations

native-country  hours-per-week
0                     Cambodia             737
1                       Canada            4438
2                        China            2628
3                     Columbia            2220
4                         Cuba            3655
5           Dominican-Republic            2833
6                      Ecuador            1086
7                  El-Salvador            3700
8                      England            3653
9                       France            1252
10                     Germany            5345
11                      Greece            1283
12                   Guatemala            2471
13                       Haiti            1561
14          Holand-Netherlands              40
15                    Honduras             455
16                        Hong             778
17                     Hungary             463
18                       India            4153
19                        Iran            1851
20                     Ireland            1018
21                       Italy            2871
22                     Jamaica            3086
23                       Japan            2594
24                        Laos             686
25                      Mexico           24754
26                   Nicaragua            1190
27  Outlying-US(Guam-USVI-etc)             586
28                        Peru            1077
29                 Philippines            7445
30                      Poland            2199
31                    Portugal            1430
32                 Puerto-Rico            4201
33                    Scotland             455
34                       South            3157
35                      Taiwan            1715
36                    Thailand             806
37             Trinadad&Tobago             670
38               United-States         1126866
39                     Vietnam            2447
40                  Yugoslavia             713